### ISSUE TITLE
RuleTester deeply merges default properties, including arrays (meaning specific elements are replaced, rather than the whole array)

### ISSUE THREAD

**Issue by G-Rath** [eslint/eslint#14728]
**Tell us about your environment**

Environment Info:

Node version: v14.16.0
npm version: v7.14.0
Local ESLint version: v7.29.0 (Currently used)
Global ESLint version: Not found
Operating System: linux 4.4.0-19041-Microsoft

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

Default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const { RuleTester } = require('eslint');

const TestRule = {
  name: __filename,
  create(context) {
    console.log(context.settings);

    return {};
  },
};

const defaultSettings = {
  hookNames: ['beforeAll', 'beforeEach', 'afterAll', 'afterEach'],
};

const ruleTester = new RuleTester({ settings: defaultSettings });

ruleTester.run('my-rule', TestRule, {
  valid: [{ code: '', settings: { hookNames: ['beforeEachTest'] } }],
  invalid: [],
});
```

```bash
node file.js
```

**What did you expect to happen?**

For the `settings` property provided in the test to ~completely~ override the default set via the `RuleTester` config constructor parameter; in particular, I'd expect it to *not* merge arrays.

While I imagine this could be useful in some situations (specifically with `env` maybe?), the docs don't mention this as being the behavior, and explicitly refer to the constructor config as used for *default* values:

> The RuleTester constructor accepts an optional object argument, which can be used to specify defaults for your test cases

To me this makes it very surprising behaviour to *merge* arrays instead of override.

**What actually happened? Please copy-paste the actual, raw output from ESLint.**

The two `hookNames` arrays are merged, resulting in the first element in the default array being replaced by the first element in the test-level version of the array:

```
console.log
  {
    describeBlocks: [],
    hookNames: [ 'beforeEachTest', 'beforeEach', 'afterAll', 'afterEach' ]
  }

  at getOption (src/settings.ts:25:11)
      at Array.forEach (<anonymous>)
```

**Steps to reproduce this issue:**

See "What did you do?" section above. Steps are:

1. `npm init -y && npm i -D eslint`
1. Copy code into `file.js`
1. `node file.js`

Output will be:

```
{
  hookNames: [ 'beforeEachTest', 'beforeEach', 'afterAll', 'afterEach' ]
}
```

**Are you willing to submit a pull request to fix this bug?**

Yup :)

**Comment by mdjermanovic** [comment]
`RuleTester` uses `lodash.merge`, which indeed merges arrays instead of overwriting them.

I agree with you that this is surprising for a general-purpose merging. However, it matches how our custom merging of eslintrc configs works with the `settings` property ([here](https://github.com/eslint/eslintrc/blob/7c43d7784e39cf0b7b102af64f703cade11252bb/lib/config-array/config-array.js#L304)).

For example, if you make these configuration files:

```js
// ------------- extend-me.js ------------------

module.exports = {
  settings: { a: [1, 2, 3] } 
};

// ------------- .eslintrc.js ------------------

module.exports = {
  extends: ["./extend-me.js"],
  settings: { a: [4] } 
};
```

calculated config will be `settings: { a: [4, 2, 3] }`

Given that in this case `RuleTester`  correctly reflects the behavior of ESLint's config merging, we probably shouldn't change it.

**Comment by G-Rath** [comment]
Well, that's interesting behaviour... 

I agree that it'd be best to have `RuleTester` match the `eslintrc` logic.

To be honest, I'm just trying to wrap my head around it; to me this feels like it effectively makes arrays in `settings` unusable, as I can't really think of any way you could work around this transparently to the user.

Do you have any thoughts on proposing to change the general logic? I figure that'd probably at least require an RFC, but it might be considered too breaking a change if it's been the behavior for a while. A alternative could be to introduce a rule-based syntax similar to what [`webpack-merge`](https://github.com/survivejs/webpack-merge) has (but not necessarily as complex).

Or maybe I'm missing something that makes this a very useful behaviour 😅🤔 (which is entirely possible)

**Comment by mdjermanovic** [comment]
I think the current behavior makes sense only when each position in the array has a specific meaning, but that should be rare and can be avoided by using objects instead of arrays.

`eslintrc` logic is now  frozen since we're working on the new config system (https://github.com/eslint/eslint/issues/13481). The new config will have a different merging logic, and the idea is to overwrite arrays (https://github.com/eslint/eslint/pull/14321#discussion_r627726368)

**Comment by nzakas** [comment]
@mdjermanovic should we close this now?

**Comment by G-Rath** [comment]
@nzakas if you're open to it, I think it would be good to discuss how supporting both the old and new config system would work.

It's awesome that this logic has been changed in the new system, but it raises concerns about expectations for plugin authors to support both systems (as I assume there will be a deprecation period, where both will be supported for at least one major version, which means at least one year) given there's no way to change the behavior in the old system and the difference can be pretty confusing (and impossible to work around without changing the API)?

Happy to be told I'm overthinking this, and if this is something you've already got documented somewhere I'm happy to be linked to that :)

**Comment by nzakas** [comment]
@G-Rath please add a comment on https://github.com/eslint/eslint/issues/13481. I won’t be revisiting the new config system until after ESLint v8.0.0 is out, but if you leave a comment there I’ll be sure to look into it more. I think we likely won’t know how big of a deal this is until we have a prototype to play with.

**Comment by mdjermanovic** [comment]
Closing since this works as intended with the current config system, and the discussion about the new config system will be continued in https://github.com/eslint/eslint/issues/13481.
